### DISCUSSION TITLE
Interactive brokers hard coded into Nautilus itself

### DISCUSSION THREAD

**Discussion by paulhwatson** [nautechsystems/nautilus_trader#1273] [Q&A] [answered]
Hi there. New to this.

I have nearly completed an adapter for IG Index but I have a couple of questions.

1. At the bottom of the TradingNodeBuilder class  in the 'build_exec_clients' method at the bottom of the file, there are a couple of lines of code specifically  dealing with the Interactive Brokers platform using a Portfolio 'set_specific_venue' method. I am unsure why this code is there.
2. I am asking this because my LiveExecutionClient adapter fails when handling the OrderFilled event. 'ExecEngine: Cannot handle order fill: no account found for ....' This error is raised by the ExecEngine '_handle_order_fill' method. The account variable is None despite me setting an account_id.

Am I missing something here. Should I be putting a platform specific piece of code into TradingNodeBuilder as well?

Many thanks
Paul

**Comment by cjdsellers** [comment, marked as answer]
Hi @paulhwatson 

Thanks for your interest and I'm glad to hear about this progress you're making!

So this code is a temporary work around due to the fact that Interactive Brokers can handle many venues, so is not a straight forward integration like for instance Binance, where the venue `BINANCE` will match the account, instrument provider and client IDs. Interactive Brokers can also require some routing setup.

If you look through `Portfolio` and follow the references to `self._venue` you'll see its referred to when fetching accounts, instruments, mainly with code like this:
```cython
account = self._cache.account_for_venue(self._venue or instrument.id.venue)
```

I've had a quick look and it seems that IG Index can also handle multiple venues, so I would suggest you also have to call `set_specific_venue` in a similar way to IB. An unfortunate limitation here is that there can only be one of these set per trader instance, which I hope isn't an issue for you.

Please let me know if any of this doesn't make sense, also @rsmb7z wrote the Interactive Brokers adapater and may have more specific information regarding this too :pray: .

**Reply by rsmb7z** [reply]
@paulhwatson 
The hard-coded lines you referring in TradingNodeBuilder are so it doesn't require the user to `set_specific_venue`. However you can simply set the venue for portfolio same as what you are using within IG adapter and it would behave as expected.

```
node.portfolio.set_specific_venue(IG_VENUE)
node.run()
```
